### ISSUE TITLE
NSInvalidArgumentException: no object at index 0 in section at index 0

### ISSUE THREAD

**Issue by mokagio** [wordpress-mobile/WordPress-iOS#18629]
**Sentry Url**: https://sentry.io/share/issue/8535b55b370f4513a6000d2ac2f6c458/
**User Count**: 1
**Count**: 1
**First Release**: 19.8.0.3
**First Seen**: 2022-05-16T09:29:18Z
**Last Seen**: 2022-05-16T09:29:18Z
**24 Hours**: 1
**30 Days**: 1
**LogID:** 91941FB3-2C68-4A21-919F-DD42DA724AB1

```
NSInvalidArgumentException: no object at index 0 in section at index 0
  File "Classes/ViewRelated/Blog/Blog Dashboard/Cards/Posts/PostsCardViewModel.swift", line 105, in PostsCardViewModel.postAt
  File "DashboardPostsListCardCell.swift", line 141, in DashboardPostsListCardCell.tableView
  File "<compiler-generated>", line null, in DashboardPostsListCardCell.tableView
  File "<compiler-generated>", line 140, in DashboardPostsListCardCell.tableView
  File "main.swift", line 7, in main
...
(19 additional frame(s) were not displayed)
```

---

Pretty sure the issue is with this `object(at:)`:

https://github.com/wordpress-mobile/WordPress-iOS/blob/bbaaac080de26d2c94a1b865bd88066f509d04b9/WordPress/Classes/ViewRelated/Blog/Blog%20Dashboard/Cards/Posts/PostsCardViewModel.swift#L103-L106

Might it be appropriate to return an `Optional<Post>`?. Or, if the assumption is that there'll always be something here, chase down how the app ended up in this inconsistent state.

**Comment by alpavanoglu** [comment]
As @mokagio posted the crash happens on access to `fetchedResultsController.object(at: indexPath)` when indexPath is (0, 0).
[This PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/18537) might be the culprit.

**Comment by hassaanelgarem** [comment]
I've looked into the crash, and I was able to notice a few things:
- Firstly, this looks like an edge case due to the low number of occurrences. (Only 1 for now)
- Secondly, from sentry's breadcrumbs, I think the user manually published a post right before the crash. And then tapped on a post in the dashboard's card.

From these observations, I have a theory as to what scenario caused the crash. 
1. The user had one draft.
2. The user opened up the draft in the editor and published it.
3. Because publishing is an asynchronous task, the post was still being displayed on the dashboard card as a draft.
4. The user then tapped that post again.
5. Before we reached `PostsCardViewModel.postAt` the post got published and removed from the `fetchController`'s fetched objects.

This scenario is similar to the scenario happening in #18637. Where we're displaying a post that should not be there, and eventually will be removed.

I can think of a few ways to tackle this:
1. If the post is being published, disable selection of that cell
2. Keep the displayed posts saved in an atomic local variable inside `PostsCardViewModel`. And depend on it in `PostsCardViewModel.postAt` instead of directly depending on the `fetchController`.

@leandroalonso what do you think?

**Comment by leandroalonso** [comment]
@hassaanelgarem were you able to reproduce the issue?

I think that sounds reasonable but how we would know that the post is being published if we don't have the reference to the object itself? 😬

**Comment by hassaanelgarem** [comment]
@leandroalonso 
> were you able to reproduce the issue?

I was able to reproduce this by tweaking the code a bit. I delayed the execution of the logic inside `DashboardPostsListCardCell.didSelectRowAt`.
```swift
func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    DispatchQueue.main.asyncAfter(deadline: .now() + 10) { // Add delay
        guard let post = self.viewModel?.postAt(indexPath), // This line causes the crash
              let viewController = self.viewController else {
            return
        }
        .... // rest of logic
    }
}
```

> how we would know that the post is being published if we don't have the reference to the object itself?

If the post is being published, we still show it under the drafts card, which means we still have a reference to it (more accurately, the fetch controller does). And we actually display that it's being published. My first suggestion was to disable selection during that state. But that won't really work, because, in order to check if the post is being published, we'll need to call `PostsCardViewModel.postAt` and we'll end up in the same place.

<img src="https://user-images.githubusercontent.com/25306722/169176815-5fe6c23c-e64c-4f88-8bc3-1f476ead3371.png" width=300> 

Maybe we should go with Gio's suggestion to return an optional `Post` from `PostsCardViewModel.postAt`. This will fix the crash. But it will leave one edge case:
- The user has two drafts ("Draft 1", "Draft 2")
- They publish "Draft 1" (index 0)
- Before it's published, they tap on it again
- It gets published and removed from the card
- We open "Draft 2" for the user as it now is in index 0, while the user originally tapped "Draft 1"
